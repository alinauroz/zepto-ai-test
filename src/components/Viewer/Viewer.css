.table {
    width: 100%;
    overflow: hidden;
    border-collapse: collapse;
    box-shadow: 0px 1px var(--light-grey), 0px -1px var(--light-grey);
    border-radius: 3px;
}

.table th, .table td {
    padding: 10px;
    border-top: 1px solid var(--light-grey);
    border-bottom: 1px solid var(--light-grey);
}

.table th {
    border-bottom: 2px solid var(--light-grey);
}

.table td:first-child, .table th:first-child {
    border-left: 1px solid var(--light-grey);
    padding-left: 60px;
}
.table td:last-child, .table th:last-child {
    border-right: 1px solid var(--light-grey);
    padding-right: 60px;
}