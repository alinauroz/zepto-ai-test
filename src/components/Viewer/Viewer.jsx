import React from 'react';
import uuid from 'react-uuid';
import PropTypes from 'prop-types';
import './Viewer.css';

/**
 * Viewer renders data array as a table. Pass an array of data with toView array of
 * objects. Custom styles and className are also supported. Default styles can be
 * overriden either using tableStyles prop or className prop.
 *
 * @component
 * @example
 *
 * const toView = [{value: "Name", key: "name"}, {value: "Age", key: "age"}]
 * const data = [{name: 'Ali', age: 22}]
 * const tableStyle = {}
 * const className = "people-viewer"
 * const widths = [40, 60]
 *
 * return (
 *  <Viewer
 *   toView = {toView}
 *   data = {data}
 *   tableStyle = {tableStyle}
 *   className = {className}
 *   width = {widths}
 *  />
 * )
 */

function Viewer({
  toView,
  data,
  tableStyle,
  className,
  widths,
}) {
  return (
    <table className={`table ${className}`} style={tableStyle}>
      <thead data-testid="table-heading">
        <tr>
          {
              toView.map((heading, index) => (
                <th
                  style={widths[index] ? { width: `${widths[index]}%` } : {}}
                  key={uuid()}
                >
                  {heading.value}
                </th>
              ))
          }
        </tr>
      </thead>
      <tbody>
        {
            data.map((row) => (
              <tr key={uuid()}>
                {
                    toView.map((heading) => <td key={uuid()}>{row[heading.key]}</td>)
                }
              </tr>
            ))
        }
      </tbody>
    </table>
  );
}

Viewer.propTypes = {
  toView: PropTypes.arrayOf(PropTypes.shape({
    value: PropTypes.string,
    key: PropTypes.string,
  })),
  data: PropTypes.arrayOf(PropTypes.objectOf(PropTypes.any)),
  tableStyle: PropTypes.objectOf(PropTypes.any),
  className: PropTypes.string,
  widths: PropTypes.arrayOf(PropTypes.number),
};

Viewer.defaultProps = {
  toView: [],
  data: [],
  tableStyle: {},
  className: '',
  widths: [],
};

export default Viewer;
