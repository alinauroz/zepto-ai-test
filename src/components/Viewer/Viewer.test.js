import React from 'react';
import { render } from '@testing-library/react';
import Viewer from './Viewer';

const sampleData = [
  {
    color: 'red',
    hex: '#f00',
  },
];

const sampleToView = [
  { value: 'Color', key: 'color' },
  { value: 'Hex Code', key: 'hex' },
];

it('should take a snapshot', () => {
  const { asFragment } = render(<Viewer data={sampleData} toView={sampleToView} />);
  expect(asFragment(<Viewer />)).toMatchSnapshot();
});
