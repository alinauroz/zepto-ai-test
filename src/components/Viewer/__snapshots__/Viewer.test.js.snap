// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <table
    class="table "
  >
    <thead
      data-testid="table-heading"
    >
      <tr>
        <th>
          Color
        </th>
        <th>
          Hex Code
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          red
        </td>
        <td>
          #f00
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
